horovod/spark/common/util.py

```diff
@@ -0,0 +1,546 @@
+    train_data_path = store.get_train_data_path(dataset_idx)
+    validation_data_path = store.get_val_data_path(dataset_idx)
+
+    if not store.exists(train_data_path):
+        raise ValueError("{} path does not exist in the store".format(train_data_path))
+
+    train_data = pq.ParquetDataset(train_data_path, filesystem=store.get_filesystem())
+    schema = train_data.schema.to_arrow_schema()
+    train_rows, total_byte_size = _get_dataset_info(train_data, 'training', train_data_path)
+
+    val_rows = 0
+    if store.exists(validation_data_path):
+        val_data = pq.ParquetDataset(validation_data_path, filesystem=store.get_filesystem())
+        val_rows, _ = _get_dataset_info(val_data, 'validation', validation_data_path)
+
+    schema_cols = feature_columns + label_columns
+    if sample_weight_col:
+        schema_cols.append(sample_weight_col)
+
+    metadata = {}
+    for col in schema_cols:
+        col_schema = schema.field_by_name(col)
+        col_info = {
+            'spark_data_type': pyarrow_to_spark_data_type(col_schema.type),
+            'is_sparse_vector_only': False,
+            'shape': None,  # Only used by SparseVector columns
+            'intermediate_format': constants.NOCHANGE,
+            'max_size': None  # Only used by SparseVector columns
+        }
+        metadata[col] = col_info
+
+    avg_row_size = total_byte_size / train_rows
+    return train_rows, val_rows, metadata, avg_row_size
+
+
+def _train_val_split(df, metadata, validation):
+    train_df = df
+    val_df = None
+    validation_ratio = 0.0
+
+    if isinstance(validation, float) and validation > 0:
+        train_df, val_df = train_df.randomSplit([1.0 - validation, validation])
+        validation_ratio = validation
+    elif isinstance(validation, str):
+        bool_dtype = metadata[validation]['spark_data_type'] == BooleanType
+        val_df = train_df.filter(
+            f.col(validation) if bool_dtype else f.col(validation) > 0).drop(validation)
+        train_df = train_df.filter(
+            ~f.col(validation) if bool_dtype else f.col(validation) == 0).drop(validation)
+
+        # Approximate ratio of validation data to training data for proportionate scale
+        # of partitions
+        timeout_ms = 1000
+        confidence = 0.90
+        train_rows = train_df.rdd.countApprox(timeout=timeout_ms, confidence=confidence)
+        val_rows = val_df.rdd.countApprox(timeout=timeout_ms, confidence=confidence)
+        validation_ratio = val_rows / (val_rows + train_rows)
+    elif validation:
+        raise ValueError('Unrecognized validation type: {}'.format(type(validation)))
+
+    return train_df, val_df, validation_ratio
+
+
+@contextlib.contextmanager
+def prepare_data(num_processes, store, df, label_columns, feature_columns,
+                 validation=None, sample_weight_col=None,
+                 partitions_per_process=10, verbose=False):
```

Comment:
maybe here `verbose` use `int` for consistent with user settings.

Reply:
Done.